Daily test of wq-example.py from coffea documentation

David-Simonetti-ND opened the pull request:
This change downloads the current wq-example.py as shown on the coffea documentation for the work_queue executor and runs it as the test. This is so that if anything changes in coffea and breaks the example, we will know when the test runs.

David-Simonetti-ND commented:
coffea-job current fails because wq-example.py from the coffea documentation currently does not work.

btovar commented:
David, it should work with cctools 4.1 and with the current master.  Investigating...

dthain commented:
@David-Simonetti-ND Ok, the CI is working as designed now -- see the merge commit.
However, now it is detecting a new problem: it looks like the factory from one run is stomping on another..

David-Simonetti-ND commented:
Fixed the error where the two factories were messing with each other. For whatever reason, the operating system was not cleaning up the wq-factory-%uid folder, so I added lines to kill the previous worker and remove the folder.

btovar commented:
@David-Simonetti-ND The `wq-factory-%uid` directory is persistent by design.  If you have a use case where it needs to be removed, should be add an option for that?

David-Simonetti-ND commented:
@btovar I think this might be more of an obscure bug than a new feature we need. The error reported when the factories were conflicting is shown below. It occurred after the first script ran successfully and after the second script began to run.
<img width="604" alt="Screen Shot 2022-01-28 at 9 36 08 AM" src="https://user-images.githubusercontent.com/31227025/151565952-154c68be-7c2a-421a-89b5-1abeac379a11.png">
So for whatever reason, some executable was still running the work_queue_worker binary after the first factory cleaned up and this caused the second factory to be unable to cp work_queue_worker to the wq-factory-uid directory. 
I even tried to explicitly kill the previous work_queue_worker that was created and tried to sleep for five seconds to give time for whatever executable was running to terminate but neither method worked. The only method that worked was to force remove the directory and have the second factory create everything from scratch.
The weird part about this is that I can not recreate the error on my ccl workstation. This occurance only seems to happen on the Github test server.
Perhaps the first factory created a second worker at some point and this worker was unable to be cleaned up properly?

David-Simonetti-ND commented:
I created a branch without manually starting work_queue_workers and let the factories spin up their own workers, but the problem still occurred.
I included a line to run ps aux after the first program completed successfully, and low and behold there is a still a work_queue_worker process running after the first program exits. 
<img width="931" alt="Screen Shot 2022-01-28 at 11 02 19 AM" src="https://user-images.githubusercontent.com/31227025/151580366-5c31c4a2-59b3-4579-9dbf-b38ba2ef6db4.png">
It looks like poncho_package_run might be keeping the worker alive? I am perplexed at this point.

dthain commented:
Aha, so it's really:

application -> work_queue_factory -> poncho_package_run -> work_queue_worker

So, I think what we need to ask of each component (by looking at the code) is:
- Does it kill its child process before exiting?
- Does it then wait for the child to actually complete before exiting?

dthain commented:
@David-Simonetti-ND what's the status on this one?
